assets/js/modules/analytics-4/datastore/custom-dimensions-gathering-data.js

```diff
@@ -0,0 +1,396 @@
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import invariant from 'invariant';
+
+/**
+ * Internal dependencies
+ */
+import API from 'googlesitekit-api';
+import Data from 'googlesitekit-data';
+import { CORE_USER } from '../../../googlesitekit/datastore/user/constants';
+import { createFetchStore } from '../../../googlesitekit/data/create-fetch-store';
+import { createReducer } from '../../../googlesitekit/data/create-reducer';
+import { getDateString } from '../../../util';
+import { CUSTOM_DIMENSION_DEFINITIONS, MODULES_ANALYTICS_4 } from './constants';
+
+const { createRegistryControl, createRegistrySelector } = Data;
+
+const RECEIVE_CUSTOM_DIMENSION_GATHERING_DATA =
+	'RECEIVE_CUSTOM_DIMENSION_GATHERING_DATA';
+const RECEIVE_CUSTOM_DIMENSION_DATA_AVAILABLE_ON_LOAD =
+	'RECEIVE_CUSTOM_DIMENSION_DATA_AVAILABLE_ON_LOAD';
+const WAIT_FOR_CUSTOM_DIMENSION_DATA_AVAILABILITY_STATE =
+	'WAIT_FOR_CUSTOM_DIMENSION_DATA_AVAILABILITY_STATE';
+
+const fetchSaveCustomDimensionDataAvailableStateStore = createFetchStore( {
+	baseName: 'saveCustomDimensionDataAvailableState',
+	controlCallback: ( { customDimension } ) =>
+		API.set( 'modules', 'analytics-4', 'custom-dimension-data-available', {
+			customDimension,
+		} ),
+	argsToParams: ( customDimension ) => ( { customDimension } ),
+	validateParams: ( { customDimension } ) => {
+		invariant(
+			'string' === typeof customDimension,
+			'customDimension must be a string.'
+		);
+	},
+} );
+
+/*
+ * The initial state shape is as follows:
+ *
+ * {
+ *   customDimensionsGatheringData: {
+ *     googlesitekit_post_date: undefined,
+ *     googlesitekit_post_author: undefined,
+ *     // etc.
+ *   },
+ *   customDimensionsDataAvailableOnLoad: {
+ *     googlesitekit_post_date: false,
+ *     googlesitekit_post_author: true,
+ *     // etc.
+ *   }
+ * }
+ */
+const baseInitialState = {
+	customDimensionsGatheringData: Object.keys(
+		CUSTOM_DIMENSION_DEFINITIONS
+	).reduce( ( initialStateSlice, customDimension ) => {
+		initialStateSlice[ customDimension ] = undefined;
+		return initialStateSlice;
+	}, {} ),
+
+	customDimensionsDataAvailableOnLoad:
+		global._googlesitekitModulesData?.[ 'analytics-4' ]
+			?.customDimensionsDataAvailable,
+};
+
+const baseActions = {
+	/**
+	 * Receives gathering data state for a custom dimension.
+	 *
+	 * @since n.e.x.t
+	 * @private
+	 *
+	 * @param {string}  customDimension Custom dimension slug.
+	 * @param {boolean} gatheringData   Gathering data.
+	 * @return {Object} Redux-style action.
+	 */
+	receiveIsCustomDimensionGatheringData( customDimension, gatheringData ) {
```

Comment:
Note that I've named this `receiveIsCustomDimensionGatheringData()` rather than `receiveIsGatheringCustomDimensionData()`, with similar choices made throughout, for the sake of consistency and to avoid breaking up the `GatheringData` term for easier searching.

Reply:
Also, please note I've used `customDimension` throughout to indicate the custom dimension "slug", rather than `customDimensionName` as per the IB.

This is also for consistency with the existing codebase e.g.:

https://github.com/google/site-kit-wp/blob/61a101f8cbeb60040d5111009b410065a7be67b9/assets/js/modules/analytics-4/utils/custom-dimensions.js#L30